protocol/protocol.go

```diff
@@ -44,3 +44,28 @@ type Protocol interface {
 	// Enveloped values are assumed to be TStructs.
 	DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error)
 }
+
+// EnvelopeAgnosticProtocol defines a specific way for a Thrift value to be
+// encoded or decoded, additionally being able to decode requests without prior
+// knowledge of whether the request is enveloped.
+type EnvelopeAgnosticProtocol interface {
```

Comment:
Might be worth also adding a comment about upcasting Protocol to this interface here.